pywemo/subscribe.py

```diff
@@ -35,32 +73,124 @@ def get_ip_address(host='1.2.3.4'):
+      device. The virtual device receives requests to change its state from
+      other Wemo devices on the network. When a Wemo device is configured to
+      change the state of the virutal device via a long press rule the following
+      sequence occurs:
+
+      1. The Wemo device will attempt to locate the virtual device on the
+      network. This is handled by the pywemo.ssdp.DiscoveryResponder class. See
+      the documentation there for more information about this step.
+
+      2. The Wemo device will fetch /setup.xml from do_GET to learn of the
+      virtual device details.
+
+      3. The Wemo device will subscribe to BinaryState notifications from the
+      virtual device. The virtual device does not send any BinaryState
+      notifications, but this step seems to be necessary before the next step
+      can happen. This step is implemented by the do_SUBSCRIBE method.
+
+      4. When a person presses the button on the Wemo for 2 seconds a long press
+      rule is triggered. If the long press rule is configured with an action
+      for the virtual device, the Wemo device will then call the do_POST method
+      to update the BinaryState of the virtual device. This doesn't actually
+      update any state, rather the virtual device then delivers the event
+      notification to any event listeners configured to receive events from the
+      pywemo SubscriptionRegistry. The event type for a long press action is
+      EVENT_TYPE_LONG_PRESS.
+    """
+
+    def do_NOTIFY(self):  # pylint: disable=invalid-name
         """Handle subscription responses received from devices."""
         sender_ip, _ = self.client_address
         outer = self.server.outer
         device = outer.devices.get(sender_ip)
-        content_len = int(self.headers.get('content-length', 0))
-        data = self.rfile.read(content_len)
         if device is None:
             LOG.warning('Received event for unregistered device %s', sender_ip)
         else:
-            # trim garbage from end, if any
-            data = data.decode("UTF-8").split("\n\n")[0]
-            doc = cElementTree.fromstring(data)
+            doc = self._get_xml_from_http_body()
             for propnode in doc.findall('./{0}property'.format(NS)):
                 for property_ in list(propnode):
                     text = property_.text
                     outer.event(device, property_.tag, text)
 
-        self.send_response(200)
-        self.send_header('Content-Type', 'text/html')
-        self.send_header('Content-Length', len(SUCCESS))
+        self._send_response(200, RESPONSE_SUCCESS)
+
+    def do_GET(self):  # pylint: disable=invalid-name
+        """Handle GET requests for a Virtual WeMo device."""
+        if self.path.endswith("/setup.xml"):
+            self._send_response(
+                200, VIRTUAL_SETUP_XML, content_type="text/xml"
+            )
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
```

Comment:
When the `eventservice.xml` is requested from the virtual device, it will return 404, right?  So the firmware on the actual wemo devices won't get the "known" actions from the `basicevent1` service then.  Since you've tested this and it works, I guess you know that it doesn't matter.  We could potentially return dummy XML data to the physical device to declare the `SetBinaryState(BinaryState, Duration, EndAction, UDN, CountdownEndTime, deviceCurrentTime)` action.  A little more management, but maybe it would be safer in case some wemo firmware versions do care.  What do you think?  Worth it or just cross that bridge if some devices have issues?

Reply:
I'm erring toward giving the Wemo less data. The more data that gets provided to the device, the more chances there are of needing to change the data in the future.

Reply:
Right.  I guess I'm just a little surprised that it worked, since technically the wemo doesn't know that the `SetBinaryState` action even exists at all!  But I guess they assume it would, since they'd expect the receiver to be a Wemo anyway.  I'm ok to leave this as-is, especially if it works.  But we should file this away in the back of our brains if there are weird issues that some users report